eventuate-local-java-cdc-connector-polling/src/main/java/io/eventuate/local/polling/PollingDao.java

```diff
@@ -125,25 +115,45 @@ public int processEvents(BinlogEntryHandler handler) {
             this::onInterrupted,
             running);
 
-    List<Object> ids = new ArrayList<>();
+    List<CompletableFuture<Object>> ids = new ArrayList<>();
 
     while (sqlRowSet.next()) {
       Object id = sqlRowSet.getObject(pk);
-      ids.add(id);
-
-      handleEvent(id, handler, sqlRowSet);
-
+      ids.add(handleEvent(id, handler, sqlRowSet));
       onEventReceived();
     }
 
-    if (ids.isEmpty()) {
+    if (!ids.isEmpty()) {
       onActivity();
+      markEventsAsProcessed(ids, pk, handler);
     }
 
     return ids.size();
   }
 
-  private void handleEvent(Object id, BinlogEntryHandler handler, SqlRowSet sqlRowSet) {
+  private void markEventsAsProcessed(List<CompletableFuture<Object>> futureIds, String pk, BinlogEntryHandler handler) {
+    List<Object> ids = futureIds
+            .stream()
+            .map(futureId -> {
+              try {
```

Comment:
Also, how many places in the CDC repeat this code pattern: cf.get() wrapped in a try/catch that throws. Define a static helper function,. CompleteFutureUtil.get(cf)

Reply:
ok, will do